### DISCUSSION TITLE
Autoflush or What is a Query?

### DISCUSSION THREAD

**Discussion by billycolley** [sqlalchemy/sqlalchemy#8498] [Usage Questions] [answered]
Hello,

I'm trying to get clarification about how autoflush works with Sessions. The documentation states the following:
`
autoflush[¶](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.params.autoflush) –

When True, all query operations will issue a [Session.flush()](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.flush) call to this Session before proceeding. This is a convenience feature so that [Session.flush()](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.flush) need not be called repeatedly in order for database queries to retrieve results. It’s typical that autoflush is used in conjunction with autocommit=False. In this scenario, explicit calls to [Session.flush()](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.flush) are rarely needed; you usually only need to call [Session.commit()](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.commit) (which flushes) to finalize changes.
`

My question is "what constitutes a query operation"? Does that mean ONLY actual `.query()` operations? Does it include insert operations as well? From my limited testing it would appear not, as I can add an insert to a session using session.add, but it isn't available until I either `.commit()` or use a `.query()` operation (isn't available if I look directly at the db). To be clear, this isn't a problem I'm having necessarily, I'm just looking for clarification as a database query could be a select query to retrieve items, or an action query to update/insert, but that doesn't SEEM to be what's happening here. Maybe I'm just misinterpreting what I'm seeing. 

Thank you for any help on this matter.

Billy Colley

**Comment by zzzeek** [comment, marked as answer]
"all query operations" in an historic sense means whenever you call a method on `Query` that will actually talk to the database.  This includes `.all()`, `.one()`, `.first()`, `iter(query)`, `update()`, `delete()`.       the ORM in its legacy mode of use, which is what that sentence was written in mind for, does not have any "insert operation" outside of the flush itself.

the document you refer to above has a seealso right there, [click that](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#session-flushing).

**Reply by billycolley** [reply]
Thank you for the quick answer! I can't believe I missed the `See Also`...

Again, thank you for the clarification!
